### DISCUSSION TITLE
Cannot override Success Behavior at template-level

### DISCUSSION THREAD

**Discussion by chrismlusk** [solspace/craft-freeform#465] [Questions & Answers] [answered]
**Describe the bug or issue you're experiencing**

[According to the docs](https://docs.solspace.com/craft/freeform/v3/overview/form-builder.html#settings-tabs), the Success Behavior setting can be overridden at template-level:

> Success Behavior - set how you'd like the success return of this form to be handled. **May also be overrided at template-level.**

But I am unable to figure out how. I can't find anything in the docs that explains how.

**Steps to reproduce**

I've tried some version of the following with both `successBehavior` and `successBehaviour` properties, and trying to set the value to either `no-effect`, `null`, `false`, or an empty string. Nothing appears to work.

```
craft.freeform.form('demo', {
  successBehavior: 'no-effect', 
})

craft.freeform.form('demo').render({
  successBehavior: 'no-effect',
})

craft.freeform.form('demo').render({
  customAttributes: {
    successBehavior: 'no-effect',
  },
})
```

**Craft & Plugin Info (please complete the following information):**

- Craft Version: 3.7.55.2
- Freeform Version: 3.13.19
- Freeform Edition: Pro

**Comment by kjmartens** [comment, marked as answer]
Hi @chrismlusk,

What it means when it says "can be overridden at template-level" is that the behavior itself can be overridden through other means.

For example, adding a `returnUrl` parameter in the template (when not using AJAX) will force the form to redirect to that URL instead of staying on the same page and showing success if that is what it is set in CP):

```
{{ form.renderTag({returnUrl: "contact/success"}) }}
```

If using AJAX and you have the form set to stay on the same page upon success or set to `None` with a return URL configured in form builder, it will also redirect to a URL if you add the following [JS override](https://docs.solspace.com/craft/freeform/v4/templating/ajax-forms.html#return-urls)

```
<html>
<head>
</head>
<body>

    {{ craft.freeform.form("myFormHandle").render() }}

    <script>
      // Do something on a successful ajax submit
      document.addEventListener('freeform-ajax-success', function(event) {
        // Redirect the user
        if (event.response.finished) {
          window.location.href = event.response.returnUrl;
        }
      })
      document.addEventListener("freeform-render-success", function (event) {
      // Stop the default success message rendering
        event.preventDefault();
      });
    </script>

</body>
</html>
```

If using AJAX and you have the form set to stay on the same page upon success and show a success message if you add the following [JS override](https://docs.solspace.com/craft/freeform/v4/templating/ajax-forms.html#return-urls) it'll cancel the success message:

```
<html>
<head>
</head>
<body>

    {{ craft.freeform.form("myFormHandle").render() }}

    <script>
      document.addEventListener("freeform-render-success", function (event) {
      // Stop the default success message rendering
        event.preventDefault();
      });
    </script>

</body>
</html>
```

I understand the wording may be confusing though, so we'll adjust it. 🙂

**Comment by smockensturm** [comment]
You should set this back to "No Effect" or something else backwards compatible. Or give us a config setting.

Client has no idea what any of this means and your new Return Behavior (3.13+) is breaking their forms. So we have to troubleshoot and then tell them why, through no fault of their own. Thanks.

And your docs are still confusing. That's how we end up here.

**Reply by kjmartens** [reply]
Sorry for the trouble @smockensturm. Would you be able to elaborate on the issue you're experiencing? Are you saying it would be beneficial to have a setting to set the default behavior for the _Success Behavior_ dropdown? 🙂 

In the meantime, the most backwards-compatible option (aside from "No Effect") would be **Reload Form with Success Message**. It should behave similarly to pre-3.13 versions.

**Comment by smockensturm** [comment]
> Sorry for the trouble @smockensturm. Would you be able to elaborate on the issue you're experiencing? Are you saying it would > be beneficial to have a setting to set the default behavior for the Success Behavior dropdown? 🙂

Yes.

> In the meantime, the most backwards-compatible option (aside from "No Effect") would be Reload Form with Success Message. It should behave similarly to pre-3.13 versions.

I know. That's why we want to configure the default behavior. :)

**Comment by iainhenderson** [comment]
I am using Freeform Pro 3.13.29.

When adding new forms the "Success behaviour" defaults to "Use return URL" with an empty value. This default means when a user successfully submits a form no success message is shown.

The default should either be "No effect" for backwards compatibility. Or should be configurable via template or by setting a default behaviour in the freeform global settings.
